rosidl_cli/rosidl_cli/command/generate/helpers.py

```diff
@@ -0,0 +1,141 @@
+    """
+    return list({
+        f'{package_name_from_include_path(path)}:{path}'
+        for include_path in map(pathlib.Path, include_paths)
+        for path in include_path.resolve().glob('**/*.idl')
+    })
+
+
+def idl_tuples_from_interface_files(interface_files):
+    """
+    Express ROS interface definition file paths as IDL tuples.
+
+    An IDL tuple is a relative path prefixed by by an absolute path against
+    which to resolve it followed by a colon ':'. This function then applies
+    the following logic:
+    - If a given path follows this pattern, it is passed through.
+    - If a given path is prefixed by a relative path, it is resolved
+      relative to the current working directory.
+    - If a given path has no prefixes, the current working directory is
+      used as prefix.
+    """
+    idl_tuples = []
+    for path in interface_files:
+        path_as_string = str(path)
+        if ':' not in path_as_string:
+            prefix = pathlib.Path.cwd()
+            stem = path
+        else:
+            prefix, _, stem = path_as_string.rpartition(':')
+            prefix = pathlib.Path(prefix).resolve()
+        stem = pathlib.Path(stem)
+        if stem.is_absolute():
+            raise ValueError('Interface definition file path '
+                             f'{stem} cannot be absolute')
+        idl_tuples.append(f'{prefix}:{stem.as_posix()}')
+    return idl_tuples
+
+
+@contextlib.contextmanager
+def legacy_generator_arguments_file(
+    *,
+    package_name,
+    interface_files,
+    include_paths,
+    templates_path,
+    output_path
+):
+    """
+    Generate a temporary rosidl generator arguments file.
+
+    :param package_name: Name of the ROS package for which to generate code
+    :param interface_files: Relative paths to ROS interface definition files,
+      optionally prefixed by another absolute or relative path followed by
+      a colon ':'. The former relative paths will be used as a prototype to
+      namespace generated code (if applicable).
+    :param include_paths: Paths where ROS package dependencies' interface
+      definition files may be found
+    :param templates_path: Path to the templates directory for the
+      generator script this arguments are for
+    :param output_path: Path to the output directory for generated code
+    """
+    idl_tuples = idl_tuples_from_interface_files(interface_files)
+    interface_dependencies = dependencies_from_include_paths(include_paths)
+    output_path = pathlib.Path(output_path).resolve()
+    templates_path = pathlib.Path(templates_path).resolve()
+    with tempfile.NamedTemporaryFile(mode='w') as tmp:
+        tmp.write(json.dumps({
+            'package_name': package_name,
+            'output_dir': str(output_path),
+            'template_dir': str(templates_path),
+            'idl_tuples': idl_tuples,
+            'ros_interface_dependencies': interface_dependencies,
+            'target_dependencies': []
```

Comment:
What's `target_dependencies` used for? Is it ok that it's empty? The arguments file from the CMake pipeline has a lot of stuff


```console
$ cat build/actionlib_msgs/rosidl_generator_c__arguments.json
{
  "package_name": "actionlib_msgs",
...
  "target_dependencies": [
    "/home/sloretz/install/rosidl_generator_c/share/rosidl_generator_c/cmake/../../../lib/rosidl_generator_c/rosidl_generator_c",
    "/home/sloretz/install/rosidl_generator_c/share/rosidl_generator_c/cmake/../../../lib/python3.8/site-packages/rosidl_generator_c/__init__.py",
    "/home/sloretz/install/rosidl_generator_c/share/rosidl_generator_c/cmake/../resource/action__type_support.h.em",
    "/home/sloretz/install/rosidl_generator_c/share/rosidl_generator_c/cmake/../resource/idl.h.em",
    "/home/sloretz/install/rosidl_generator_c/share/rosidl_generator_c/cmake/../resource/idl__functions.c.em",
    "/home/sloretz/install/rosidl_generator_c/share/rosidl_generator_c/cmake/../resource/idl__functions.h.em",
    "/home/sloretz/install/rosidl_generator_c/share/rosidl_generator_c/cmake/../resource/idl__struct.h.em",
    "/home/sloretz/install/rosidl_generator_c/share/rosidl_generator_c/cmake/../resource/idl__type_support.h.em",
    "/home/sloretz/install/rosidl_generator_c/share/rosidl_generator_c/cmake/../resource/msg__functions.c.em",
    "/home/sloretz/install/rosidl_generator_c/share/rosidl_generator_c/cmake/../resource/msg__functions.h.em",
    "/home/sloretz/install/rosidl_generator_c/share/rosidl_generator_c/cmake/../resource/msg__struct.h.em",
    "/home/sloretz/install/rosidl_generator_c/share/rosidl_generator_c/cmake/../resource/msg__type_support.h.em",
    "/home/sloretz/install/rosidl_generator_c/share/rosidl_generator_c/cmake/../resource/srv__type_support.h.em",
    "/home/sloretz/build/actionlib_msgs/rosidl_adapter/actionlib_msgs/msg/GoalID.idl",
    "/home/sloretz/build/actionlib_msgs/rosidl_adapter/actionlib_msgs/msg/GoalStatus.idl",
    "/home/sloretz/build/actionlib_msgs/rosidl_adapter/actionlib_msgs/msg/GoalStatusArray.idl",
    "/home/sloretz/install/builtin_interfaces/share/builtin_interfaces/msg/Duration.idl",
    "/home/sloretz/install/builtin_interfaces/share/builtin_interfaces/msg/Time.idl",
    "/home/sloretz/install/std_msgs/share/std_msgs/msg/Bool.idl",
    "/home/sloretz/install/std_msgs/share/std_msgs/msg/Byte.idl",
    "/home/sloretz/install/std_msgs/share/std_msgs/msg/ByteMultiArray.idl",
    "/home/sloretz/install/std_msgs/share/std_msgs/msg/Char.idl",
    "/home/sloretz/install/std_msgs/share/std_msgs/msg/ColorRGBA.idl",
    "/home/sloretz/install/std_msgs/share/std_msgs/msg/Empty.idl",
    "/home/sloretz/install/std_msgs/share/std_msgs/msg/Float32.idl",
    "/home/sloretz/install/std_msgs/share/std_msgs/msg/Float32MultiArray.idl",
    "/home/sloretz/install/std_msgs/share/std_msgs/msg/Float64.idl",
    "/home/sloretz/install/std_msgs/share/std_msgs/msg/Float64MultiArray.idl",
    "/home/sloretz/install/std_msgs/share/std_msgs/msg/Header.idl",
    "/home/sloretz/install/std_msgs/share/std_msgs/msg/Int16.idl",
    "/home/sloretz/install/std_msgs/share/std_msgs/msg/Int16MultiArray.idl",
    "/home/sloretz/install/std_msgs/share/std_msgs/msg/Int32.idl",
    "/home/sloretz/install/std_msgs/share/std_msgs/msg/Int32MultiArray.idl",
    "/home/sloretz/install/std_msgs/share/std_msgs/msg/Int64.idl",
    "/home/sloretz/install/std_msgs/share/std_msgs/msg/Int64MultiArray.idl",
    "/home/sloretz/install/std_msgs/share/std_msgs/msg/Int8.idl",
    "/home/sloretz/install/std_msgs/share/std_msgs/msg/Int8MultiArray.idl",
    "/home/sloretz/install/std_msgs/share/std_msgs/msg/MultiArrayDimension.idl",
    "/home/sloretz/install/std_msgs/share/std_msgs/msg/MultiArrayLayout.idl",
    "/home/sloretz/install/std_msgs/share/std_msgs/msg/String.idl",
    "/home/sloretz/install/std_msgs/share/std_msgs/msg/UInt16.idl",
    "/home/sloretz/install/std_msgs/share/std_msgs/msg/UInt16MultiArray.idl",
    "/home/sloretz/install/std_msgs/share/std_msgs/msg/UInt32.idl",
    "/home/sloretz/install/std_msgs/share/std_msgs/msg/UInt32MultiArray.idl",
    "/home/sloretz/install/std_msgs/share/std_msgs/msg/UInt64.idl",
    "/home/sloretz/install/std_msgs/share/std_msgs/msg/UInt64MultiArray.idl",
    "/home/sloretz/install/std_msgs/share/std_msgs/msg/UInt8.idl",
    "/home/sloretz/install/std_msgs/share/std_msgs/msg/UInt8MultiArray.idl"
  ]
}
```

Reply:
IIUC those are for file caching (see [here](https://github.com/ros2/rosidl/blob/57befb87e165fc660809b30fba1a28e587263a4b/rosidl_cmake/rosidl_cmake/__init__.py#L62) and [here](https://github.com/ros2/rosidl/blob/57befb87e165fc660809b30fba1a28e587263a4b/rosidl_cmake/rosidl_cmake/__init__.py#L166-L173)). 

It's a bit of a nice to have, so I skipped it for now. But I can expose it as an argument so we can add that later on.

Reply:
See https://github.com/ros2/rosidl/pull/567/commits/b581585ab48e924efe9c78445301767bd87451fe.